dsc/workflows/base/base_workflow.py

```diff
@@ -0,0 +1,199 @@
+            will be uploaded
+            output_queue: The SQS output queue used for retrieving result messages from
+            the workflow's submissions.
+        """
+        self.workflow_name: str = workflow_name
+        self.submission_system: str = submission_system
+        self.email_recipients: list[str] = email_recipients
+        self.metadata_mapping: dict = metadata_mapping
+        self.s3_bucket: str = s3_bucket
+        self.s3_prefix: str | None = s3_prefix
+        self.collection_handle: str = collection_handle
+        self.output_queue: str = output_queue
+
+    @final  # noqa: B027
+    def run(self) -> None:
+        """Run workflow to submit items to  the DSpace Submission Service.
+
+        PLANNED CODE:
+
+        sqs_client = SQSClient()
+        for item_submission in self.item_submissions_iter():
+            item_submission.upload_dspace_metadata(
+                self.s3_bucket, item_submission.metadata_keyname
+            )
+            sqs_client.send_submission_message(
+                item_submission.item_identifier,
+                self.workflow_name,
+                self.output_queue,
+                self.submission_system,
+                self.collection_handle,
+                item_submission.metadata_uri,
+                item_submission.bitstream_uris,
+            )
+        """
+
+    @final
+    def item_submissions_iter(self) -> Iterator[ItemSubmission]:
+        """Generate a batch of item submissions for the DSpace Submission Service.
+
+        MUST NOT be overridden by workflow subclasses.
+        """
+        for item_metadata in self.batch_metadata_iter():
+            item_identifier = self.get_item_identifier(item_metadata)
+            logger.info(f"Processing submission for '{item_identifier}'")
+            metadata_keyname = f"{self.s3_prefix}/{item_identifier}_metadata.json"
+            dspace_metadata = self.create_dspace_metadata(item_metadata)
+            if self.validate_dspace_metadata(dspace_metadata):
+                item_submission = ItemSubmission(
+                    dspace_metadata=dspace_metadata,
+                    bitstream_uris=self.get_bitstream_uris(item_identifier),
+                    metadata_keyname=metadata_keyname,
+                )
+                yield item_submission
+
+    @abstractmethod
+    def batch_metadata_iter(self) -> Iterator[dict[str, Any]]:
+        """Iterate through batch metadata to yield item metadata.
+
+        MUST be overridden by workflow subclasses.
+        """
+
+    @abstractmethod
+    def get_item_identifier(self, item_metadata: Any) -> str:  # noqa: ANN401
+        """Get identifier for an item submission according to the workflow subclass.
+
+        MUST be overridden by workflow subclasses.
+
+        Args:
+            item_metadata: The item metadata from which the item identifier is extracted.
+        """
+
+    @final
+    def create_dspace_metadata(self, item_metadata: dict[str, Any]) -> dict[str, Any]:
```

Comment:
If this pattern is flexible enough for all workflows, I think this could be a real strength of this application: requiring basically a "raw" item dictionary + workflow mapping = final dspace metadata.  If this is truly the only transformation logic needed, across all workflows, that would be pretty great!

Reply:
I am fairly confident that it will!